File: stack/typescript/valid-parentheses.ts
//https://leetcode.com/problems/valid-parentheses/
//
//TODO: I don't like line 11
const isValid = (s: string): boolean => {
  const dict = {
    ")": "(",
    "]": "[",
    "}": "{",
  }

  const stack: string[] = [];

  s.split('').forEach(c => {
    if (stack[stack.length - 1] === dict[c] && dict[c] !== undefined) {
      stack.pop()
    } else {
      stack.push(c)
    }
  })

  return !stack.length
};


console.log(isValid("()[]}{"))

File: stack/go/valid-parentheses.go
// https://leetcode.com/problems/valid-parentheses
// TODO: solve again
package main

import (
	"fmt"
)

func main() {
	test := "]"
	fmt.Println(isValid(test))
}

func isValid(s string) bool {
	openingParenthesesMap := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}

	closingParenthesesMap := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	sliceAsPseudoStack := []rune{}

	for _, v := range s {
		_, isOpening := openingParenthesesMap[v]

		if isOpening {
			sliceAsPseudoStack = append(sliceAsPseudoStack, v)
			continue
		}

		if len(sliceAsPseudoStack) <= 0 {
			return false
		}

		lastItem := sliceAsPseudoStack[len(sliceAsPseudoStack)-1]
		opening, _ := closingParenthesesMap[v]

		if lastItem != opening {
			return false
		}
		sliceAsPseudoStack = sliceAsPseudoStack[:len(sliceAsPseudoStack)-1]
	}

	return len(sliceAsPseudoStack) == 0
}
